js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+        toJS: {
+            value(): FieldDataMapping<FinalFields> {
+                const data = {};
+                for (const [fieldName, value] of Object.entries(this._data)) {
+                    data[fieldName] = this._model.fields[fieldName].toJS(value);
+                }
+                return data as FieldDataMapping<FinalFields>;
+            },
+        },
+        isEqual: {
+            value(
+                record: ViewModelInterface<FieldsMapping, { [fieldName: string]: any }>
+            ): boolean {
+                if (record._model !== this._model) {
+                    return false;
+                }
+                if (!isEqual(record._assignedFields, this._assignedFields)) {
+                    return false;
+                }
+                for (const fieldName of this._assignedFields) {
+                    const field = this._model.fields[fieldName];
+                    if (!field.isEqual(this[fieldName], record[fieldName])) {
+                        return false;
+                    }
+                }
+                return true;
+            },
+        },
+        clone: {
+            value(
+                fieldNames?: string[]
+            ): ViewModelInterface<FinalFields, FinalFields, PkFieldType, PkValueType> {
+                const missingFieldNames = fieldNames
+                    ? fieldNames.filter(fieldName => !this._assignedFields.includes(fieldName))
+                    : [];
+                if (fieldNames && missingFieldNames.length > 0) {
+                    throw new Error(
+                        `Can't clone ${this._model.name} with fields ${fieldNames.join(
+                            ', '
+                        )} as only these fields are set: ${this._assignedFields.join(
+                            ', '
+                        )}. Missing fields: ${missingFieldNames.join(', ')}`
+                    );
+                }
+                if (!fieldNames) {
+                    fieldNames = this._assignedFields as string[];
+                }
+
+                const data = {};
+                for (const fieldName of fieldNames) {
+                    // TODO: Unclear to me if this needs to call a method on the Field on not. Revisit this.
+                    data[fieldName] = this[fieldName];
+                }
+
+                // Always clone primary keys
+                const pkFieldNames = this._model.pkFieldNames;
+                pkFieldNames.forEach(name => (data[name] = this[name]));
+
+                // I don't know how to type this, error is:
+                // TS2322: Type 'ViewModel' is not assignable to type 'this'.
+                // 'ViewModel' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'ViewModel'.
+                // I could type return as `ViewModel` but then usages of it will be wrong (eg. clone() will be from a more general ViewModel to a specific implementation)
```

Comment:
"eg. clone() will be from a more general ViewModel to a specific implementation"

do you mean

"we want clone() to return a specific implementation type instead of a general ViewModel"?

(Was a bit unclear what you were saying)

Reply:
I've removed this entirely - seemed to be remnant of original class implementation and error no longer exists